File: ADAS/CARLA-Simulator/pid_control/PID.cpp
#include "PID.hpp"

double mapPIDtoServo(double pid_value, double min_angle, double max_angle) {

    // Clamp pid_value to [-1, 1]
    if (pid_value > 1.0) pid_value = 1.0;
    if (pid_value < -1.0) pid_value = -1.0;

    double servo_angle = (pid_value + 1.0) * (max_angle - min_angle) / 2.0 + min_angle;

    return (servo_angle);
}

static double clampSymmetric(double value, double limit) {

    if (limit <= 0.0)
        return 0.0;
	else if (value > limit)
        return limit;
    else if (value < -limit)
        return -limit;

    return value;
}

PID::PID(double kp, double ki, double kd)
    : _kp(kp),
      _ki(ki),
      _kd(kd),
      _prevError(0.0),
      _integral(0.0),
      _integralLimit(1.0),
      _outputLimit(1.0) {}

PID::~PID() {}

double PID::update(double target, double current, double dt) {

    if (dt <= 0.0)
        return 0.0;

    const double error = target - current;
    const double proportional = _kp * error;

    _integral += error * dt;
    _integral = clampSymmetric(_integral, _integralLimit);

    const double integral = _ki * _integral;
    const double derivative = _kd * ((error - _prevError) / dt);

    _prevError = error;

    const double output = proportional + integral + derivative;
    return clampSymmetric(output, _outputLimit);
}

void PID::reset() {
    _prevError = 0.0;
    _integral = 0.0;
}

void PID::setIntegralLimit(double limit) {
    _integralLimit = std::abs(limit);
    _integral = clampSymmetric(_integral, _integralLimit);
}

void PID::setOutputLimit(double limit) {
    _outputLimit = std::abs(limit);
}

File: ADAS/CARLA-Simulator/pid_control/pid.py
import numpy as np

def clamp_symmetric(value, limit):
    if limit <= 0.0:
        return 0.0
    elif value > limit:
        return limit
    elif value < -limit:
        return -limit
    return value


def map_pid_to_servo(pid_value, min_angle, max_angle):
    # Clamp to [-1, 1]
    pid_value = max(-1.0, min(1.0, pid_value))

    servo_angle = (pid_value + 1.0) * (max_angle - min_angle) / 2.0 + min_angle
    return servo_angle


class PID:

    def __init__(self, kp, ki, kd):
        self.kp = kp
        self.ki = ki
        self.kd = kd

        self.prev_error = 0.0
        self.integral = 0.0

        self.integral_limit = 1.0
        self.output_limit = 1.0

    def update(self, target, current, dt):

        if dt <= 0.0:
            return 0.0

        error = target - current

        proportional = self.kp * error

        self.integral += error * dt
        self.integral = clamp_symmetric(self.integral, self.integral_limit)

        integral = self.ki * self.integral

        derivative = self.kd * ((error - self.prev_error) / dt)

        self.prev_error = error

        output = proportional + integral + derivative

        return clamp_symmetric(output, self.output_limit)

    def reset(self):
        self.prev_error = 0.0
        self.integral = 0.0

    def set_integral_limit(self, limit):
        self.integral_limit = abs(limit)
        self.integral = clamp_symmetric(self.integral, self.integral_limit)

    def set_output_limit(self, limit):
        self.output_limit = abs(limit)